src/common/io_exerciser/RadosIo.cc

```diff
@@ -0,0 +1,279 @@
+#include "RadosIo.h"
+
+#include "DataGenerator.h"
+
+using RadosIo = ceph::io_exerciser::RadosIo;
+
+RadosIo::RadosIo(librados::Rados& rados,
+        boost::asio::io_context& asio,
+        const std::string& pool,
+        const std::string& oid,
+        uint64_t block_size,
+        int seed,
+	int threads,
+        ceph::mutex& lock,
+        ceph::condition_variable& cond) :
+  Model(oid, block_size),
+  rados(rados),
+  asio(asio),
+  om(std::make_unique<ObjectModel>(oid, block_size, seed)),
+  db(data_generation::DataGenerator::create_generator(
+      data_generation::GenerationType::HeaderedSeededRandom, *om)),
+  pool(pool),
+  threads(threads),
+  lock(lock),
+  cond(cond),
+  outstanding_io(0)
+{
+  int rc;
+  rc = rados.ioctx_create(pool.c_str(), io);
+  ceph_assert(rc == 0);
+  allow_ec_overwrites(true);
+}
+
+RadosIo::~RadosIo()
+{
+}
+
+void RadosIo::start_io()
+{
+  std::lock_guard l(lock);
+  outstanding_io++;
+}
+
+void RadosIo::finish_io()
+{
+  std::lock_guard l(lock);
+  ceph_assert(outstanding_io > 0);
+  outstanding_io--;
+  cond.notify_all();
+}
+
+void RadosIo::wait_for_io(int count)
+{
+  std::unique_lock l(lock);
+  while (outstanding_io > count) {
+    cond.wait(l);
+  }
+}
+
+void RadosIo::allow_ec_overwrites(bool allow)
+{
+  int rc;
+  bufferlist inbl, outbl;
+  std::string cmdstr =
```

Comment:
I have not updated this with fmt::format because I am planning on adding more json commands in another update separate to this merge request and would prefer to deal with this section in that request. I do agree with you this could be easier to read and I would like to spend the time when I create that PR to update the json creation and make it common code between the commands rather than how it is created here, along with investigating creating them through json libraries included in the ceph source already.